CesiumGltf/include/CesiumGltf/SharedAssetDepot.h

```diff
@@ -0,0 +1,356 @@
+#pragma once
+
+#include "CesiumGltf/ImageCesium.h"
+
+#include <CesiumAsync/AsyncSystem.h>
+#include <CesiumAsync/Future.h>
+#include <CesiumAsync/IAssetAccessor.h>
+#include <CesiumAsync/IAssetResponse.h>
+#include <CesiumUtility/IntrusivePointer.h>
+#include <CesiumUtility/ReferenceCounted.h>
+
+#include <cstddef>
+#include <forward_list>
+#include <functional>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <unordered_set>
+#include <variant>
+
+namespace CesiumGltf {
+
+template <typename AssetType> class SharedAsset;
+template <typename AssetType> class SingleAssetDepot;
+
+/**
+ * Contains the current state of an asset within the SharedAssetDepot.
+ */
+template <typename AssetType>
+class AssetContainer
+    : public CesiumUtility::ReferenceCounted<AssetContainer<AssetType>, true> {
+public:
+  std::atomic<std::int32_t> counter;
+  std::string assetId;
+  AssetType asset;
+
+  // Pointer to this container's parent so we know how to clean it up.
+  SingleAssetDepot<AssetType>* parent;
+
+  AssetContainer(
+      std::string assetId_,
+      AssetType&& asset_,
+      SingleAssetDepot<AssetType>* parent_)
+      : counter(0),
+        assetId(assetId_),
+        asset(std::move(asset_)),
+        parent(parent_) {}
+  SharedAsset<AssetType> toRef() { return SharedAsset(this); }
+};
+
+/**
+ * An asset that may or may not be stored in the SharedAssetDepot and shared
+ * across multiple tiles. This can either be an ImageCesium itself or a
+ * reference-counted pointer to an entry in the SharedAssetDepot. You should
+ * always treat it as if it was an ImageCesium, not a smart pointer.
+ * @tparam AssetType The type of the asset that we're getting a reference to.
+ */
+template <typename AssetType> class SharedAsset {
+public:
+  SharedAsset(AssetType&& asset)
+      : contents(new AssetContainer<AssetType>(
+            std::string(),
+            std::forward<AssetType>(asset),
```

Comment:
I noticed this while reviewing CesiumGS/cesium-unreal#1521 and wanted to mention it before I forget.

This `std::forward` doesn't make sense (nor does the one below). `std::forward` should only be used with forwarding references, which look like `T&&` where T is a template parameter _on that particular function_. Because your `AssetType` in this case is a parameter of the class, not the function/constructor, it's not a forwarding reference. It's just a regular r-value reference, and so `std::move` should be used instead of `std::forward`. The behavior is the same with either option, but the use of std::forward is misleading.

The constructor below is more interesting. It's really unusual to use `std::forward` with something declared `AssetType&`, and I wasn't sure offhand what it would do. But apparently what it does in this case is cast to an r-value reference. So the user calls a constructor, passing it an l-value, and the constructor moves it away as if it were an r-value. Really confusing.

I think what you want here is two constructors. One takes `AssetType&&` and uses `std::move`. The other takes `const AssetType&`, effectively promising not to modify what was passed in. But then how do you give it to `AssetContainer`, which only accepts the asset as an r-value reference? You can create an r-value from a const l-value by copying it:

```cpp
new AssetContainer<AssetType>(
  std::string(),
  AssetType(asset),
  nullptr)
```

Or, alternatively, you can add an `AssetContainer` constructor taking a const reference, too.

It's also possible to use one constructor with a forwarding reference instead of two. That would look like this:

```cpp
template <typename T>
SharedAsset(T&& asset)
    : contents(new AssetContainer<AssetType>(
          std::string(),
          std::forward<T>(asset),
          nullptr)) {}
```

Notice the use of `T` instead of `AssetType`. The downside to this is the compiler errors will be fairly horrific when there's a problem, such as when `T` and `AssetType` are not the same type.